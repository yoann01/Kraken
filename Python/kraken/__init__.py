"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
