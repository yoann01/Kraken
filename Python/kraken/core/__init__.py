"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
