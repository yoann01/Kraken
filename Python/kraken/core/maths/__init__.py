"""Kraken - math module."""

from math_object import MathObject
from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo
from mat33 import Mat33
from mat44 import Mat44
from rotation_order import RotationOrder
from color import Color
from constants import DEG_TO_RAD, RAD_TO_DEG, PI


def Math_radToDeg(val):
    """Converts radians to degrees.

    Args:
        val (float): value to convert to degrees.

    Returns:
        float: Value in degrees.

    """

    return val * RAD_TO_DEG

def Math_degToRad(val):
    """Converts degrees to radians.

    Args:
        val (float): value to convert to radians.

    Returns:
        float: Value in radians.

    """

    return val * DEG_TO_RAD


def decodeValue(jsonData):
    """Returns a constructed math value based on the provided json data.

    Args:
        jsondata (dict): The JSON data to use to decode into a Math value.

    Returns:
        object: The constructed math value

    """

    if type(jsonData) is not dict:
        return jsonData

    if '__mathObjectClass__' not in jsonData:
        raise Exception("Invalid JSON data for constructing value:" + str(jsonData))

    if jsonData['__mathObjectClass__'] == 'Vec2':
        val = Vec2()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec3':
        val = Vec3()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec4':
        val = Vec4()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Euler':
        val = Euler()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Quat':
        val = Quat()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Xfo':
        val = Xfo()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat33':
        val = Mat33()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat44':
        val = Mat44()
        val.jsonDecode(jsonData, decodeValue)
    else:
        raise Exception("Unsupported Math type:" + jsonData['__mathObjectClass__'])

    return val
