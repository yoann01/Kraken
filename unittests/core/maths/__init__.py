
import unittest

import test_vec2
import test_vec3
import test_vec4
import test_color
import test_euler
import test_mat33
import test_mat44
import test_quat
import test_rotation_order
import test_xfo

loadVec2Suite = test_vec2.suite()
loadVec3Suite = test_vec3.suite()
loadVec4Suite = test_vec4.suite()
loadColorSuite = test_color.suite()
loadEulerSuite = test_euler.suite()
loadMat33Suite = test_mat33.suite()
loadMat44Suite = test_mat44.suite()
loadQuatSuite = test_quat.suite()
loadRotationOrderSuite = test_rotation_order.suite()
loadXfoSuite = test_xfo.suite()


def suite():
    suites = [
        loadVec2Suite,
        loadVec3Suite,
        loadVec4Suite,
        loadColorSuite,
        loadEulerSuite,
        loadMat33Suite,
        loadMat44Suite,
        loadQuatSuite,
        loadRotationOrderSuite,
        loadXfoSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
