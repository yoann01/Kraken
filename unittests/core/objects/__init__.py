
import unittest

from attributes import suite as attributeSuite
from components import suite as componentSuite
from constraints import suite as constraintSuite
from operators import suite as operatorSuite

from test_component_group import suite as componentGroupSuite
from test_container import suite as containerSuite
from test_control import suite as controlSuite
from test_ctrlSpace import suite as ctrlSpaceSuite
from test_curve import suite as curveSuite
from test_hierarchy_group import suite as hierarchyGroupSuite
from test_joint import suite as jointSuite
from test_layer import suite as layerSuite
from test_locator import suite as locatorSuite
from test_object_3d import suite as object3DSuite
from test_rig import suite as rigSuite
from test_scene_item import suite as sceneItemSuite
from test_transform import suite as transformSuite

loadAttributeSuite = attributeSuite()
loadComponentSuite = componentSuite()
loadConstraintSuite = constraintSuite()
loadOperatorSuite = operatorSuite()

loadComponentGroupSuite = componentGroupSuite()
loadContainerSuite = containerSuite()
loadControlSuite = controlSuite()
loadCtrlSpaceSuite = ctrlSpaceSuite()
loadCurveSuite = curveSuite()
loadHierarchyGroupSuite = hierarchyGroupSuite()
loadJointSuite = jointSuite()
loadLayerSuite = layerSuite()
loadLocatorSuite = locatorSuite()
loadObject3DSuite = object3DSuite()
loadRigSuite = rigSuite()
loadSceneItemSuite = sceneItemSuite()
loadTransformSuite = transformSuite()


def suite():
    suites = [
        loadAttributeSuite,
        loadComponentSuite,
        loadConstraintSuite,
        loadOperatorSuite,
        loadContainerSuite,
        loadControlSuite,
        loadCtrlSpaceSuite,
        loadCurveSuite,
        loadHierarchyGroupSuite,
        loadJointSuite,
        loadLayerSuite,
        loadLocatorSuite,
        loadObject3DSuite,
        loadRigSuite,
        loadSceneItemSuite,
        loadTransformSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
