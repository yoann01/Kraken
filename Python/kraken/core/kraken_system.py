"""KrakenSystem - objects.kraken_core module.

Classes:
KrakenSystem - Class for constructing the Fabric Engine Core client.

"""

import logging
import os
import sys
import json
import importlib
from collections import OrderedDict

import FabricEngine.Core

# import kraken
from kraken.core.profiler import Profiler
from kraken.plugins import getFabricClient

from kraken.log import getLogger
from kraken.log.utils import fabricCallback

logger = getLogger('kraken')


class KrakenSystem(object):
    """The KrakenSystem is a singleton object used to provide an interface with
    the FabricEngine Core and RTVal system."""

    __instance = None

    def __init__(self):
        """Initializes the Kraken System object."""

        super(KrakenSystem, self).__init__()

        self.client = None
        self.typeDescs = None
        self.registeredTypes = None
        self.loadedExtensions = []

        self.registeredConfigs = OrderedDict()
        self.registeredComponents = OrderedDict()
        # self.moduleImportManager = ModuleImportManager()


    def loadCoreClient(self):
        """Loads the Fabric Engine Core Client"""

        if self.client is None:
            Profiler.getInstance().push("loadCoreClient")

            client = getFabricClient()
            if client is None:
                options = {
                    'reportCallback': fabricCallback,
                    'guarded': True
                }

                client = FabricEngine.Core.createClient(options)

            self.client = client

            self.loadExtension('Math')
            self.loadExtension('Kraken')
            self.loadExtension('KrakenForCanvas')

            Profiler.getInstance().pop()

    def getCoreClient(self):
        """Returns the Fabric Engine Core Client owned by the KrakenSystem

        Returns:
            object: The Fabric Engine Core Client

        """

        if self.client is None:
            self.loadCoreClient()

        return self.client

    def loadExtension(self, extension):
        """Loads the given extension and updates the registeredTypes cache.

        Args:
            extension (str): The name of the extension to load.

        """

        if extension not in self.loadedExtensions:
            Profiler.getInstance().push("loadExtension:" + extension)
            self.client.loadExtension(extension)
            self.registeredTypes = self.client.RT.types
            self.typeDescs = self.client.RT.getRegisteredTypes()
            # Cache the loaded extension so that we aviod refreshing the typeDescs cache(costly)
            self.loadedExtensions.append(extension)
            Profiler.getInstance().pop()


    # ==============
    # RTVal Methods
    # ==============
    def convertFromRTVal(self, target, RTTypeName=None):
        """Generates an RTVal object based on the simple type of target
        and passes target to constructor. Converts a property of an RTVal object
        to its own pytholn RTVal object


        Args:
            target (RTVal): The RTVal object or property to cast
            RTTypeName (str): The type of RTVal to convert to

        Returns:
            RTVal: The RTVal object

        """

        self.loadCoreClient()

        if RTTypeName is None:
            RTTypeName = target.type('String').getSimpleType()
        rtValType = getattr(self.client.RT.types, RTTypeName)
        pythonRTVal = rtValType(target)

        return pythonRTVal

    def constructRTVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        self.loadCoreClient()
        klType = getattr(self.registeredTypes, dataType)

        if defaultValue is not None:
            if hasattr(defaultValue, '_rtval'):
                return defaultValue._rtval

            typeDesc = self.typeDescs[dataType]
            if 'members' in typeDesc:
                try:
                    value = klType.create()
                except:
                    try:
                        return klType()
                    except Exception as e:
                        raise Exception("Error constructing RTVal:" + dataType)

                for i in xrange(0, len(typeDesc['members'])):
                    memberName = typeDesc['members'][i]['name']
                    memberType = typeDesc['members'][i]['type']
                    if memberName in defaultValue:
                        setattr(value, memberName, self.constructRTVal(memberType, getattr(defaultValue, memberName)))

                return value

            else:
                return klType(defaultValue)
        else:
            try:
                return klType.create()
            except:
                try:
                    return klType()
                except Exception as e:
                    raise Exception("Error constructing RTVal:" + dataType)

    def rtVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        return self.constructRTVal(dataType, defaultValue)

    def isRTVal(self, value):
        """Returns true if the given value is an RTVal.

        Args:
            value (value): value to test.

        Returns:
            bool: True if successful.

        """

        return str(type(value)) == "<type 'PyRTValObject'>"

    def getRTValTypeName(self, rtval):
        """Returns the name of the type, handling extracting the name from KL RTVals.

        Args:
            rtval (rtval): The rtval to extract the name from.

        Returns:
            bool: True if successful.

        """

        if ks.isRTVal(rtval):
            return json.loads(rtval.type("Type").jsonDesc("String").getSimpleType())['name']
        else:
            return "None"

    # ==================
    # Config Methods
    # ==================
    def registerConfig(self, configClass):
        """Registers a config Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            configClass (str): The Python class of the config

        """

        configModulePath = configClass.__module__ + "." + configClass.__name__

        self.registeredConfigs[configModulePath] = configClass

    def getConfigClass(self, className):
        """Returns the registered Python config class with the given name

        Args:
            className (str): The name of the Python config class

        Returns:
            object: The Python config class

        """

        if className not in self.registeredConfigs:
            raise Exception("Config with that class not registered:" + className)

        return self.registeredConfigs[className]

    def getConfigClassNames(self):
        """Returns the names of the registered Python config classes

        Returns:
            list: The array of config class names.

        """

        return self.registeredConfigs.keys()

    # ==================
    # Component Methods
    # ==================
    def registerComponent(self, componentClass):
        """Registers a component Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            componentClass (str): The Python class of the component

        """
        componentClassPath = componentClass.__module__ + "." + componentClass.__name__
        if componentClassPath in self.registeredComponents:
            # we allow reregistring of components because as a component's class is edited
            # it will be re-imported by python(in Maya), and the classes reregistered.
            pass

        self.registeredComponents[componentClassPath] = componentClass

    def getComponentClass(self, className):
        """Returns the registered Python component class with the given name

        Args:
            className (str): The name of the Python component class

        Returns:
            object: The Python component class

        """

        if className not in self.registeredComponents:
            raise Exception("Component with that class not registered:" + className)

        return self.registeredComponents[className]

    def getComponentClassNames(self):
        """Returns the names of the registered Python component classes

        Returns:
            list: The array of component class names.

        """

        return self.registeredComponents.keys()

    def loadComponentModules(self):
        """Loads all the component modules and configs specified in the 'KRAKEN_PATHS' environment variable.

        The kraken_components are loaded at all times.

        Returns:
            bool: True if all components loaded, else False.

        """

        for componentClassPath in self.registeredComponents:
            componentModulePath = self.registeredComponents[componentClassPath].__module__
            if componentModulePath in sys.modules:
                del(sys.modules[componentModulePath])

        self.registeredComponents = {}

        logger.info("Loading component modules...")

        def __importDirRecursive(path, parentModulePath=''):
            isSuccessful = True

            contents = os.listdir(path)
            moduleFilefound = False
            for item in contents:
                if os.path.isfile(os.path.join(path, item)):
                    if item == "__init__.py":
                        if parentModulePath == '':
                            modulePath = os.path.basename(path)

                            moduleParentFolder = os.path.split( path )[0]
                            if moduleParentFolder not in sys.path:
                                sys.path.append(moduleParentFolder)
                        else:
                            modulePath = parentModulePath + '.' + os.path.basename(path)
                        moduleFilefound = True


            if moduleFilefound:
                logger.info(" " + path + ":")
                for i, item in enumerate(contents):
                    if os.path.isfile(os.path.join(path, item)):

                        # Parse all the files of given path and import python
                        # modules. The files in these folders really should be
                        # limited to components, otherwise we are loading more
                        # than modules and that is not clear.

                        # TODO: Figure out a way to limit imports to just rig
                        # component modules.

                        if item.endswith(".py") and item != "__init__.py":
                            module = modulePath + "." + item[:-3]
                            try:
                                logger.info("  " + module)
                                importlib.import_module(module)

                            except ImportError, e:
                                isSuccessful = False
                                logging.exception("Error importing '" + module)

                            except Exception, e:
                                isSuccessful = False
                                logging.exception("Error Loading Modules'" + module)

                logging.info("")


            for item in contents:
                if os.path.isdir(os.path.join(path, item)):
                    if moduleFilefound:
                        if not __importDirRecursive(os.path.join(path, item), modulePath):
                            isSuccessful = False
                    else:
                        if not __importDirRecursive(os.path.join(path, item)):
                            isSuccessful = False

            return isSuccessful


        # Find the kraken examples module in the same folder as the kraken module.
        default_component_path = os.path.normpath(os.path.join(os.environ.get('KRAKEN_PATH'), 'Python', 'kraken_components'))
        isSuccessful = __importDirRecursive(default_component_path)

        pathsVar = os.getenv('KRAKEN_PATHS')
        if pathsVar is not None:
            pathsList = pathsVar.split(os.pathsep)
            for path in pathsList:

                if path == '':
                    continue

                if not os.path.exists(path):
                    logging.info("Invalid Kraken Path: " + path)
                    continue

                if not __importDirRecursive(path):
                    isSuccessful = False

        return isSuccessful


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the KrakenSystem

        Returns:
            object: The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = KrakenSystem()

        return cls.__instance


ks = KrakenSystem.getInstance()
