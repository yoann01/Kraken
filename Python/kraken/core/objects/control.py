"""Kraken - objects.Controls.Control module.

Classes:
Control - Base Control.

"""

from kraken.core.configs.config import Config
from kraken.core.maths.constants import AXIS_NAME_TO_TUPLE_MAP
from kraken.core.maths import Euler, Quat, Vec3, Xfo
from kraken.core.maths import Math_degToRad
from kraken.core.objects.curve import Curve
from kraken.core.objects.ctrlSpace import CtrlSpace


class Control(Curve):
    """Base Control object."""

    def __init__(self, name, parent=None, shape="null", scale=1.0, flags=None, metaData=None):
        """Initializes control object.

        Args:
            name (str): Name of the constraint.
            parent (Object): Parent object of this object.

        """

        super(Control, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
        self.shape = shape

        self.setShape(shape)

        self.scalePoints(Vec3(scale,scale,scale))

    # ==============
    # Shape Methods
    # ==============
    def getShape(self):
        """Returns the shape that the control was set to.

        Returns:
            str: Name of the shape that was set.

        """

        return self.shape

    def setShape(self, shape):
        """Sets the shape of the control to the one specified.

        Args:
            shape (str): the desired shape of the control.

        Returns:
            bool: True if successful.

        """

        config = Config.getInstance()
        configShapes = config.getControlShapes()
        if shape not in configShapes.keys():
            raise KeyError("'" + shape + "' is not a valid shape in the loaded config of class [" + config.__class__.__name__ + "]")

        self.shape = shape
        self.setCurveData(configShapes[shape])

        return True


    # ==============
    # Align Methods
    # ==============
    def alignOnXAxis(self, negative=False):
        """Aligns the control shape on the X axis.

        Args:
            negative (bool): Whether to align the control on the negative X axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[0] < furthest:
                        furthest = eachPoint[0]

                else:

                    if eachPoint[0] > furthest:
                        furthest = eachPoint[0]


        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] += offset

        self.setCurveData(curveData)

        return True

    def alignOnYAxis(self, negative=False):
        """Aligns the control shape on the Y axis.

        Args:
            negative (bool): Whether to align the control on the negative Y axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[1] < furthest:
                        furthest = eachPoint[1]

                else:

                    if eachPoint[1] > furthest:
                        furthest = eachPoint[1]

        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[1] += offset

        self.setCurveData(curveData)

        return True

    def alignOnZAxis(self, negative=False):
        """Aligns the control shape on the Z axis.

        Args:
            negative (bool): Whether to align the control on the negative Z axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[2] < furthest:
                        furthest = eachPoint[2]

                else:

                    if eachPoint[2] > furthest:
                        furthest = eachPoint[2]

        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[2] += offset

        self.setCurveData(curveData)

        return True


    # ==============
    # Scale Methods
    # ==============
    def scalePointsOnAxis(self, scale, scaleAxis="POSX"):
        """Scales the point positions from it's center along the given axis only.

        Args:
            scale (float): scale value to apply to the points.
            scaleAxis (str): which axes to scale and by what direction

        Returns:
            bool: True if successful.

        """

        # would be great if vec3 was iterable
        axis = AXIS_NAME_TO_TUPLE_MAP.get(scaleAxis)

        if axis is None:
            raise KeyError("'" + scaleAxis + "' is not a valid axis. Valid axes are: " + ','.join(AXIS_NAME_TO_TUPLE_MAP.keys()))

        scaleList = [1.0, 1.0, 1.0]
        for i, x in enumerate(axis):
            if x != 0:
                scaleList[i] = scale * axis[i]

        return self.scalePoints(Vec3(scaleList[0], scaleList[1], scaleList[2]))


    # ==============
    # Scale Methods
    # ==============
    def scalePoints(self, scaleVec):
        """Scales the point positions from it's center.

        Args:
            scaleVec (Vec3): scale value to apply to the points.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] *= scaleVec.x
                eachPoint[1] *= scaleVec.y
                eachPoint[2] *= scaleVec.z

        self.setCurveData(curveData)

        return True


    # ===============
    # Rotate Methods
    # ===============
    def rotatePoints(self, xRot, yRot, zRot):
        """Rotates the points by the input values.

        Args:
            xRot (float): Euler x rotate value.
            yRot (float): Euler y rotate value.
            zRot (float): Euler z rotate value.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        quatRot = Quat()
        quatRot.setFromEuler(Euler(Math_degToRad(xRot), Math_degToRad(yRot),
                                   Math_degToRad(zRot)))

        for eachSubCurve in curveData:

            for eachPoint in eachSubCurve["points"]:
                pointVec = Vec3(eachPoint[0], eachPoint[1], eachPoint[2])
                rotVec = quatRot.rotateVector(pointVec)
                eachPoint[0] = rotVec.x
                eachPoint[1] = rotVec.y
                eachPoint[2] = rotVec.z

        self.setCurveData(curveData)

        return True


    # ==================
    # Translate Methods
    # ==================
    def translatePoints(self, translateVec):
        """Translates the control points.

        Args:
            translateVec (Vec3): Translation values to apply to the points.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] += translateVec.x
                eachPoint[1] += translateVec.y
                eachPoint[2] += translateVec.z

        self.setCurveData(curveData)

        return True


    # ===============
    # Helper Methods
    # ===============
    def insertCtrlSpace(self, name=None):
        """Adds a CtrlSpace object above this object

        Args:
            name (string): optional name for this CtrlSpace, default is same as
                this object

        Returns:
            object: New CtrlSpace object

        """

        if name is None:
            name = self.getName()

        newCtrlSpace = CtrlSpace(name, parent=self.getParent())
        if self.getParent() is not None:
            self.getParent().removeChild(self)

        if self.getMetaDataItem("altLocation") is not None:
            newCtrlSpace.setMetaDataItem("altLocation", self.getMetaDataItem("altLocation"))

        self.setParent(newCtrlSpace)
        newCtrlSpace.addChild(self)

        newCtrlSpace.xfo = Xfo(self.xfo)

        # To ensure that names of control spaces don't clash with controls and
        # if they do, set's the control space's name back to what it was intended
        if self.getName() == name:
            newCtrlSpace.setName(name)

        return newCtrlSpace
