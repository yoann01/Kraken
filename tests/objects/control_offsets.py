from kraken import plugins
from kraken.core.maths import *
from kraken.core.objects.control import Control


builder = plugins.getBuilder()

config = builder.getConfig()
config.setExplicitNaming(True)

nullControl = Control("NullControl", shape="null")
nullControl.rotatePoints(0, 45, 0)
nullControl.scalePoints(Vec3(3, 3, 3))
nullControl.translatePoints(Vec3(0, 1, 0.25))
nullControl.xfo.tr = Vec3(0.0, 1.0, 0.0)

builder.build(nullControl)
