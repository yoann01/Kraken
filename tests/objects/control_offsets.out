Failed to find DCC builder. Falling back to Python builder.
building: NullControl as: NullControl type: Control
buildControl:NullControl as: NullControl
connectAttribute: NullControl.implicitAttrGrp.visibility
connectAttribute: NullControl.implicitAttrGrp.shapeVisibility
